File: 3/js/p1.js
const readline = require('readline');

let counts = null;

const rl = readline.createInterface({
  input: process.stdin
});

rl.on('line', (l) => {
  if (counts === null) counts = new Array(l.length).fill(0);

  for (let i = 0; i < l.length; i++) {
    const n = l[i];

    if (n === '1') counts[i]++;
    else if (n === '0') counts[i]--;
    else throw `Unknown character (${n})`;
  }
});

rl.on('close', () => {
  let gamma = '';
  let epsilon = '';

  for (let i = 0; i < counts.length; i++) {
    if (counts[i] > 0) {
      gamma += '1';
      epsilon += '0';
    } else {
      gamma += '0';
      epsilon += '1';
    }
  }

  let gn = parseInt(gamma, 2);
  let en = parseInt(epsilon, 2);

  console.log(`Gamma: ${gamma} ${gn}`);
  console.log(`Gamma: ${epsilon} ${en}`);
  console.log(`${gn * en}`);
});

File: 3/lua/p1.lua
local counts = {}
local gamma, epsilon = '', ''

for line in io.lines() do
  for i = 1, #line do
    local n = line:sub(i, i)

    if counts[i] == nil then counts[i] = 0 end

    if n == '1' then
      counts[i] = counts[i] + 1
    elseif n == '0' then
      counts[i] = counts[i] - 1
    else
      error('Unknown character: '..n)
    end
  end
end

for _,v in ipairs(counts) do
  if v > 0 then
    gamma = gamma .. '1'
    epsilon = epsilon .. '0'
  else
    gamma = gamma .. '0'
    epsilon = epsilon .. '1'
  end
end

print(gamma, tonumber(gamma, 2))
print(epsilon, tonumber(epsilon, 2))
print(tonumber(gamma, 2) * tonumber(epsilon, 2))
